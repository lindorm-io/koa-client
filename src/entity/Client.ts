import { ClientEvent } from "../enum";
import { EntityBase, EntityCreationError, IEntity, IEntityBaseOptions, TObject } from "@lindorm-io/core";

export interface IClient extends IEntity {
  approved: boolean;
  description: string;
  extra: TObject<any>;
  name: string;
  secret: string;
}

export interface IClientOptions extends IEntityBaseOptions {
  approved?: boolean;
  description?: string;
  extra?: TObject<any>;
  name?: string;
  secret?: string;
}

export class Client extends EntityBase implements IClient {
  private _approved: boolean;
  private _description: string;
  private _extra: TObject<any>;
  private _name: string;
  private _secret: string;

  constructor(options?: IClientOptions) {
    super(options);
    this._approved = options?.approved || false;
    this._description = options?.description || null;
    this._extra = options?.extra || {};
    this._name = options?.name || null;
    this._secret = options?.secret || null;
  }

  public get approved(): boolean {
    return this._approved;
  }
  public set approved(approved: boolean) {
    this._approved = approved;
    this.addEvent(ClientEvent.APPROVED_CHANGED, { approved: this._approved });
  }

  public get description(): string {
    return this._description;
  }
  public set description(description: string) {
    this._description = description;
    this.addEvent(ClientEvent.DESCRIPTION_CHANGED, { description: this._description });
  }

  public get extra(): TObject<any> {
    return this._extra;
  }
  public set extra(object: TObject<any>) {
    this._extra = object;
    this.addEvent(ClientEvent.EXTRA_CHANGED, { extra: this._extra });
  }

  public get name(): string {
    return this._name;
  }
  public set name(name: string) {
    this._name = name;
    this.addEvent(ClientEvent.NAME_CHANGED, { name: this._name });
  }

  public get secret(): string {
    return this._secret;
  }
  public set secret(secret: string) {
    this._secret = secret;
    this.addEvent(ClientEvent.SECRET_CHANGED, { secret: this._secret });
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== ClientEvent.CREATED) continue;
      throw new EntityCreationError("Client");
    }
    this.addEvent(ClientEvent.CREATED, {
      approved: this._approved,
      description: this._description,
      extra: this._extra,
      name: this._name,
      secret: this._secret,
      created: this._created,
      updated: this._updated,
    });
  }
}
