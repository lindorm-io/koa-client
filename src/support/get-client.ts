import { Client } from "../entity";
import { IKoaClientContext } from "../types";

export interface IGetClientOptions {
  disableCache?: boolean;
  disableRepository?: boolean;
}

export const getClient = (ctx: IKoaClientContext) => async (
  clientId: string,
  options?: IGetClientOptions,
): Promise<Client> => {
  const { cache, logger, repository } = ctx;
  const { disableCache, disableRepository } = options || {};

  if (!disableCache) {
    try {
      logger.debug("find client in cache", { clientId });

      return await cache.client.find(clientId);
    } catch (err) {
      logger.warn("client not found in cache", err);

      if (disableRepository) {
        throw err;
      }
    }
  }

  if (!disableRepository) {
    try {
      logger.debug("find client in repository", { id: clientId });

      return await repository.client.find({ id: clientId });
    } catch (err) {
      logger.error("client not found in repository", err);

      throw err;
    }
  }

  throw new Error("Client not found due to invalid server settings");
};
