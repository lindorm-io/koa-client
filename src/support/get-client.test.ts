import MockDate from "mockdate";
import { CacheEntityNotFoundError } from "@lindorm-io/redis";
import { Client } from "../entity";
import { RepositoryEntityNotFoundError } from "@lindorm-io/mongo";
import { getClient } from "./get-client";
import {
  getGreyBoxCache,
  getGreyBoxRepository,
  inMemoryCache,
  inMemoryStore,
  logger,
  resetCache,
  resetStore,
} from "../test";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));

MockDate.set("2020-01-01 08:00:00.000");

describe("getClient", () => {
  let ctx: any;

  beforeEach(async () => {
    ctx = {
      cache: await getGreyBoxCache(),
      logger,
      repository: await getGreyBoxRepository(),
    };
  });

  afterEach(() => {
    resetCache();
    resetStore();
  });

  test("should resolve client from cache", async () => {
    await ctx.cache.client.create(
      new Client({
        id: "950b90fc-0d07-4712-8226-192ae18cbc7c",
      }),
    );

    await expect(getClient(ctx)("950b90fc-0d07-4712-8226-192ae18cbc7c")).resolves.toMatchSnapshot();

    expect(inMemoryCache).toMatchSnapshot();
  });

  test("should resolve client from repository", async () => {
    await ctx.repository.client.create(
      new Client({
        id: "950b90fc-0d07-4712-8226-192ae18cbc7c",
      }),
    );

    await expect(getClient(ctx)("950b90fc-0d07-4712-8226-192ae18cbc7c")).resolves.toMatchSnapshot();

    expect(inMemoryStore).toMatchSnapshot();
  });

  test("should throw CacheEntityNotFoundError", async () => {
    await expect(getClient(ctx)("950b90fc-0d07-4712-8226-192ae18cbc7c", { disableRepository: true })).rejects.toThrow(
      expect.any(CacheEntityNotFoundError),
    );
  });

  test("should throw RepositoryEntityNotFoundError", async () => {
    await expect(getClient(ctx)("950b90fc-0d07-4712-8226-192ae18cbc7c", { disableCache: true })).rejects.toThrow(
      expect.any(RepositoryEntityNotFoundError),
    );
  });
});
