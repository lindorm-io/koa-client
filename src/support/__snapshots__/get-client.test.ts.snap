// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getClient should resolve client from cache 1`] = `
Client {
  "_approved": false,
  "_created": 2020-01-01T07:00:00.000Z,
  "_description": null,
  "_events": Array [],
  "_extra": Object {},
  "_id": "950b90fc-0d07-4712-8226-192ae18cbc7c",
  "_name": null,
  "_secret": null,
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;

exports[`getClient should resolve client from cache 2`] = `
Object {
  "Client::950b90fc-0d07-4712-8226-192ae18cbc7c": Object {
    "blob": "{\\"json\\":{\\"id\\":\\"950b90fc-0d07-4712-8226-192ae18cbc7c\\",\\"version\\":\\"0\\",\\"created\\":\\"2020-01-01T07:00:00.000Z\\",\\"updated\\":\\"2020-01-01T07:00:00.000Z\\",\\"events\\":[],\\"approved\\":\\"false\\",\\"description\\":\\"null\\",\\"extra\\":{},\\"name\\":\\"null\\",\\"secret\\":\\"null\\"},\\"meta\\":{\\"id\\":\\"string\\",\\"version\\":\\"number\\",\\"created\\":\\"date\\",\\"updated\\":\\"date\\",\\"events\\":[],\\"approved\\":\\"boolean\\",\\"description\\":\\"null\\",\\"extra\\":{},\\"name\\":\\"null\\",\\"secret\\":\\"null\\"}}",
    "expiresInSeconds": null,
  },
}
`;

exports[`getClient should resolve client from repository 1`] = `
Client {
  "_approved": false,
  "_created": 2020-01-01T07:00:00.000Z,
  "_description": null,
  "_events": Array [
    Object {
      "date": 2020-01-01T07:00:00.000Z,
      "name": "client_created",
      "payload": Object {
        "approved": false,
        "created": 2020-01-01T07:00:00.000Z,
        "description": null,
        "extra": Object {},
        "name": null,
        "secret": null,
        "updated": 2020-01-01T07:00:00.000Z,
      },
    },
  ],
  "_extra": Object {},
  "_id": "950b90fc-0d07-4712-8226-192ae18cbc7c",
  "_name": null,
  "_secret": null,
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;

exports[`getClient should resolve client from repository 2`] = `
Object {
  "database_name": Object {
    "client": Object {
      "data": Array [
        Object {
          "approved": false,
          "created": 2020-01-01T07:00:00.000Z,
          "description": null,
          "events": Array [
            Object {
              "date": 2020-01-01T07:00:00.000Z,
              "name": "client_created",
              "payload": Object {
                "approved": false,
                "created": 2020-01-01T07:00:00.000Z,
                "description": null,
                "extra": Object {},
                "name": null,
                "secret": null,
                "updated": 2020-01-01T07:00:00.000Z,
              },
            },
          ],
          "extra": Object {},
          "id": "950b90fc-0d07-4712-8226-192ae18cbc7c",
          "name": null,
          "secret": null,
          "updated": 2020-01-01T07:00:00.000Z,
          "version": 0,
        },
      ],
      "indices": Array [
        Object {
          "index": Object {
            "id": 1,
          },
          "options": Object {
            "unique": true,
          },
        },
      ],
    },
  },
}
`;
